Dynamic component loading draft

vitek-karas opened the pull request:
Very first draft of the "plugin" design.

TODOs:
- [x] Short description of .deps.json functionality - to demonstrate the capabilities we'll enable by using it.
- [ ] Section about expected tooling experiences - what does it mean to build a "plugin" using .NET SDK and so on.
- [ ] Section about settings/config knobs - env. variables, roll forward settings, command line arguments  - which will have effect of component loading and which will be ignored and how.
- [ ] Section about interactions of the new ALC with existing extension points of the binder (various events on `Assembly`, `AppDomain` and `AssemblyLoadContext`, interation with other ALCs, native asset resolution extension points ...)
- [x] More details on native asset resolution
- [x] More details on satellite assembly resolution
- [x] Handling of NI images

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+
+## Problem statement
+.NET Core runtime has only limited support for dynamically loading assemblies which have additional dependencies or possibly collide with the app in any way. Out of the box only these scenarios really work:
+* Assemblies which have no additional dependencies other than those find in the app itself (`Assembly.Load` and `Assembly.LoadFrom`)
```
those find -> those found

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+
+## Problem statement
+.NET Core runtime has only limited support for dynamically loading assemblies which have additional dependencies or possibly collide with the app in any way. Out of the box only these scenarios really work:
+* Assemblies which have no additional dependencies other than those find in the app itself (`Assembly.Load` and `Assembly.LoadFrom`)
```
I do not think that `Assembly.LoadFrom` should be mentioned in the first bullet. It is mentioned in the second bullet already.

vitek-karas replied:
I was thinking about it... `Assembly.Load` only loads by assembly name, `Assembly.LoadFrom` can do the same but given a file path. But I'll remove it for simplicity.

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+* `static Assembly Assembly.LoadFileWithDependencies` - in its core similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
```
I would describe the API shape in the standard "API review" format, like:

```csharp
class Assembly
{
    public static Assembly LoadFileWithDependencies(string path);
}
```

jeffschwMSFT replied:
For unification purposes, I would recommend that this hang off of AssemblyLoadContext.  The universe of Assembly.Load* and AppDomain.* have legacy behaviors that may not be intuitive.  From a higher level strategy I think people should direct people towards ALCs.

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+
+Public API (early thinking):
+* `static Assembly Assembly.LoadFileWithDependencies` - in its core similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+* `static AssemblyLoadContext AssemblyLoadContext.CreateForAssemblyWithDependencies(string assemblyPath)` - "advanced" version which would return an ALC instance and not just the main assembly. Could have overloads for enabling unloadability.
```
Add the proposed API with the unloading flag.

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* `static Assembly Assembly.LoadFileWithDependencies` - in its core similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+* `static AssemblyLoadContext AssemblyLoadContext.CreateForAssemblyWithDependencies(string assemblyPath)` - "advanced" version which would return an ALC instance and not just the main assembly. Could have overloads for enabling unloadability.
+
+## High-level description of the solution
```
Nit: "High-level description of the solution" -> `High-level description of the implementation" ?

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to TPA and list of native search paths. These will be used to resolve any assembly binding events in the load context.
+* If the load context can't resolve an assembly bind event, it will fallback to the parent load context (the app)
```
We may consider to have an advanced API that lets you specify load context to fallback to.

vitek-karas replied:
I added that in the API section.

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* The host (the app) can be any configuration (framework dependent or self-contained). The notion of frameworks is completely ignored by this new functionality.
+* All framework dependencies of the component must be resolvable by the app - simply put, the component must use the same frameworks as the app.
+* Components can't add frameworks to the app - the app must "pre-load" all necessary frameworks.
+* By default only framework types will be shared between the app and the component. The component may chose to share more by not including dependencies in the component (`CopyLocal=false`).
```
We may want to have a separate paragraph for tooling experience. How do we expect one to build the "component".

vitek-karas replied:
Added this as a TODO for now - I'll have to do a bit more research on that.

jeffschwMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+## Problem statement
+.NET Core runtime has only limited support for dynamically loading assemblies which have additional dependencies or possibly collide with the app in any way. Out of the box only these scenarios really work:
+* Assemblies which have no additional dependencies other than those find in the app itself (`Assembly.Load` and `Assembly.LoadFrom`)
+* Assemblies which have additional dependencies in the same folder and which don't collide with anything in the app itself (`Assembly.LoadFrom`)
```
In general this does not work for the dynamic case.  The only way dependencies are found is if they are described in the main *.deps.json.

swaroop-sridhar replied:
I think this point is trying to capture the dependencies specified via `--additional-deps`

vitek-karas replied:
This section describes current behavior, basically trying to describe why a new thing is needed. 

`Assembly.LoadFrom` currently implements a mechanism to resolve dependencies which are in the same folder as the loaded assembly. It's the closest thing we have to "Dynamically load a component". It doesn't look at `.deps.json` in any way, and it loads everything into the Default context which makes it pretty hard to use.

jeffschwMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+## High-level description of the solution
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
```
Did we consider a non-*.deps.json scenario where there is just a loose set of files?  dotnet/coreclr issue 13277 has the customer scenario for that case.

vitek-karas replied:
Most of the dotnet/coreclr#13277 seems to actually want us to use `.deps.json` and the current lack of that functionalisty is the most blocking problem.

As for the case of no `.deps.json` and files in the same folder, which is also mentioned in that issue - that is actually described here:
> Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies. 

In short - if there's no `.deps.json` the ALC will effectively turn into "load next to me" ALC. It will behave as if there was a `.deps.json` listing all the files in the same folder and nothing else.

As for the NuGet problems in the issue - Our current solution is the `.deps.json` in combination with `.runtimeconfig.dev.json`. There is room for future improvements which would allow some of this configuration to be done programatically without the files on disk. But I personally haven't seen a user case which would need that, so I think we should postpone that for now.

jeffschwMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to TPA and list of native search paths. These will be used to resolve any assembly binding events in the load context.
```
+ Satellite assemblies

vitek-karas replied:
I added it as a TODO - I'll need to do some more research to fully understand our current handling of satellite assemblies.

jeffschwMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,48 @@
+* If the load context can't resolve an assembly bind event, it will fallback to the parent load context (the app)
+
+## Important implications and limitations
+* Only framework dependent components will be supported. Self-contained components will not be supported even if there was a way to produce them.
```
Is there a technical reason for this limitation?  In general the host is able to resolve self contained references just as well as framework dependent.

swaroop-sridhar replied:
This restriction is a combination of
* Inability to override certain system libraries (like CoreLib)
* Untested behavior of overriding other system libraries.
* Implementing a simple solution to start with -- with the understanding that this solution will be sufficient for mainstream plugins.

In the appmodel meeting I think @Jkotas suggested that we should start with this restriction and can allow relaxations if there is a use-case.

vitek-karas replied:
I'm just going to reword what @swaroop-sridhar  sais:
* Can can't override some assets - S.P.CoreLib, the runtime itself. So we would have to special case those.
* In fact overriding anything in `Microsoft.NETCore.App` is dangerous. Assemblies in `Microsoft.NETCore.App` can have direct references into S.P.CoreLib. And since S.P.CoreLib will be only one, allowing different versions of assemblies which depend on it may cause breaks. As far as I know we are free to introduce breaking changes to the S.P.CoreLib surface area at any time we want since only assemblies in `Microsoft.NETCore.App` can refer to it and we service that as a whole.
* The above would require us to somehow recognize anything from `Microsoft.NETCore.App`, which is actually rather tricky to do. Self-contained apps don't carry good description which assembly came from which framework. So we would porbably need a non-trivial set of hardcoded rules around this.
* For higher level frameworks this is technically possible and should not break anything really. I think it's counter intuitive in msot cases so we should make this "hard", but possible. Which it actually is. In theory I can create an app which depends only on one framework `Microsoft,NETCore.App` but carries a full copy of ASPNET framework as "self-contained". It would probably be pretty hard to produce this with current SDK, but one can craft a `.deps.json` like that if needed. - for this case I think we should wait for a specific user scenario before we go there. It might be a lot of work without a clear need right now.


All that said - it is definitely possible to include a one-of assembly from `Microsoft.NETCore.App` in the component itself and it will be used over the one from the framework. Just like it works today with apps.

There's also another issue with with SDK. For a NetStandard classlib the SDK actually ignores any attempts to make it self-contained and will always produce what is effectively a framework dependent thing. For NetCoreApp classlib, it is possible to produce a self-contained publish, but it's really weird (it gets `apphost.exe` - and I mean literally `apphost.exe`, it's not renamed, regardless of the name of the classlib). The apphost.exe is actually not even bound with the managed dll, so it shows a really bad error message.

jkotas replied:
> For NetCoreApp classlib, it is possible to produce a self-contained publish, but it's really weird

Sounds like something worth filling an SDK issue for.

vitek-karas replied:
Right. Filed:
dotnet/sdk#2550 and dotnet/sdk#2549.

sdmaclea replied:
In a sense these must be 'app dependent components" where the app is required to provide the .NET runtime and related framework.

jkotas commented:
> From a higher level strategy I think people should direct people towards ALCs.

The expert users have been directed towards ALC already.

I think there is a benefit in having the simple mainstream API to not be tangled with ALC.

jeffschwMSFT commented:
> I think there is a benefit in having the simple mainstream API to not be tangled with ALC.

In that case, I think we need to reconcile the behaviors of the other AppDomain.* to ensure they all play nicely together.

jkotas commented:
> reconcile the behaviors of the other AppDomain.* to ensure they all play nicely together.

Example what do you have in mind? The existing `Assembly.Load*` APIs have a well defined behavior that should not be affected by introducing this new thing.

jeffschwMSFT commented:
We added behaviors to AssemblyResolve event, etc. to act just like desktop.  Do we want to carry those behaviors along with this new model?  If you do an AppDomain.LoadFrom the default assembly resolve event will kick in and try to resolve a failed attempt.  How will that play with the new ALC resolve event that handles the deps.json.

jkotas commented:
> Do we want to carry those behaviors along with this new model?

Yes, I do not see why we would break the existing APIs and change them to behave differently. We may not like how these APIs behave, but that is not a good reason for breaking them.

`Assembly.LoadFrom` does what it does. It is unpredictable API in .NET Framework. It is more predictable best-effort implementation in .NET Core. It rarely does (in either .NET Core or .NET Framework) what people expect it to do. (Nit: It is `Assembly.LoadFrom` not `AppDomain.LoadFrom`.)

vitek-karas commented:
Trying to answer @jeffschwMSFT  questions:

>What other work is dependent/related to this: unifying pub and bin, unifying dotnet-cli and VS working directory on run, unloadability, etc.

I think this is independent. If this new ALC takes into consideration the `.runtimeconfig.dev.json` then it should be able to load components produced with both `dotnet build` and `dotnet publish`. It should also not depend on current directory in any way (at least as far as I can tell). There's a larger issue of how it interacts with broader environment (env. variables, command line options, app's `.runtimeconfig.json` and so on) - this is currently a TODO for this doc (see the PR description).

Unloadability is already mentioned here, and I added a new API overload which would directly enable it on the new ALC.

> I think we need to step back and put out a proposal of ALC and what all we plan on doing it together. We are adding individual pieces here and there, but I feel we need a bit more unified evolution.

Here I also think that these are almost entirely independent. Improving experience for people trying to implement their own ALC is almost completely orthogonal to us providing a stock implementation of an ALC.  
I added a new public API overload which allows specifying fallback (parent) ALC, which might be of use for the more advanced scenarios (`dotnet watch` using this ALC to load the app, which will in turn try to load a plugin would probably need this).

Similar to the other discussion in this PR - this proposal is to provide a solution to a problem for people to use without them needing to know much about ALCs or us forcing them to go learn ALCs.

> What is outstanding from this proposal that should be addressed in later ones?

In this PR description there's a list of TODOs for things I know we need to add to this document. On top of that I only think the usual applies (so I didn't list it anywhere):
* detailed technical design
* review from interested parties
* test plan
* docs
* ….

vitek-karas commented:
#### Regarding new API on `Assembly` versus `AssemblyLoadContext`
Given the feedback we've got so far on various `Assembly.Load*` methods and custom ALC implementations I think the main goal of this feature should be:
> Add a simple to use mechanism to dynamically load a component with reasonable defaults

To me this also comes with
> Don't force people onto `AssemblyLoadContext` for a typical use case

That's what the `Assembly.LoadFileWithDependencies` API is for. For people which just want sensible defaults and follow our golden paths (regarding tooling and deployment and so on). This set of users should not need to know anything about ALC.

The APIs on `AssemblyLoadContext` are for the more advanced scenarios, where it's reasonable to assume at least a high-level idea about ALC, isolation and so on.


#### Regarding interaction with existing extension points
I added an item to the TODO list to go into more details about this. But in general I agree with @jkotas, I don't see a reason to treat this new ALC in any special way. That said I'm sure I'm missing a lot of things, so if you have a specific case in mind, please let me know.

swaroop-sridhar commented:
For the TODO of "Short description of .deps.json functionality," how about

*Abilities of deps.json*
`deps.json` is a dependency manifest for dotnet apps. It enables plugins to load dependencies from locations other than the base directory (ex: from packages, platform-specific publish directories, etc.)

CoreCLR first builds a set of directories to look for binaries (called `ProbingPaths` ) based on application/plugin base directory, framework directory, runtimeconfig.dev.json, command line, servicing locations, etc.

For an application or plugin, the deps.json file specifies:
* A set of binary dependencies -- which are searched within ProbingPaths
*  Relative paths to locate them --  Relative paths within `deps.json` file can be used to locate architecture-specific dependencies within publish/package directories.
If the app depends on one or more frameworks, the deps.json files of those framework are similarly processed.

Further details about the algorithm used for processing dependencies can be found at:
https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/assembly-conflict-resolution.md

vitek-karas commented:
Thanks a lot @swaroop-sridhar - I added it into the document (only slightly modified to fir the structure).

sdmaclea commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,89 @@
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial amount of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.
```
Nit change 'amount' to 'set'

sdmaclea commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,89 @@
+* Only framework dependent components will be supported. Self-contained components will not be supported even if there was a way to produce them.
+* The host (the app) can be any configuration (framework dependent or self-contained). The notion of frameworks is completely ignored by this new functionality.
+* All framework dependencies of the component must be resolvable by the app - simply put, the component must use the same frameworks as the app.
+* Components can't add frameworks to the app - the app must "pre-load" all necessary frameworks.
```
Is this a strict requirement for all frameworks?  With the possibility of multiple frameworks this seems like an unnecessary restriction.  It might just work.  Is this just for restricting the set of supported and tested features?  

Perhaps it is reasonable to defer this until it becomes an issue.

vitek-karas replied:
There are some small issues technically, but nothing major. This is more in line of what we want to support for now. The problem is that frameworks might not expect to exist in multiple copies (for example WPF is likely to have issues if two copies are running in the same process). There's also the problem of sharing. Reasonable expectation for frameworks is that they are "shared" by all components in the app. That would require us to load the framework into the default context (even when resolving it for the component) - which comes with lot of tricky issues.

sdmaclea commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,89 @@
+* Section about settings/config knobs - env. variables, roll forward settings, command line arguments - which will have effect of component loading and which will be ignored and how.
+* Section about interactions of the new ALC with existing extension points of the binder (various events on `Assembly`, `AppDomain` and `AssemblyLoadContext`, interaction with other ALCs, native asset resolution extension points ...)
+* More details on native asset resolution
+* More details on satellite assembly resolution
```
Can we think about plug-in versioning and app compatibility.  Is there a reasonable need to provide a standardized mechanism for plugin to express app requirements.  This might be as simple as adding a deps.json section with convenience API to read it.  

If I was going to create plug-in picker, I would only want to show compatible plug-ins.

vitek-karas replied:
There are things arleady in place which might help with this. For example:
* The plugin is likely to have a dependency on a an assembly from the app (the one which defines the interface they're going to talk through).
* This dependency is in the `.deps.json` and it has a version - the version which the plugin was compiled with.
* There's existing managed API to read `.deps.json` including this information.

It's not exactly first class support, but at least something is there.

I'll add it to the list of possible improvements to add something more specific.

sdmaclea replied:
> The plugin is likely to have a dependency on a an assembly from the app

Since it will be provided by the app it will not typically be needed in the plugins deps.json.  So it might want to be declared specially.  Something like 
```json
"pluginAppDependencies" : [ "MyAppPluginModel-3.3.x" ]
```

vitek-karas replied:
Currently even if a dependency is set to not be included in the output, it will still be listed in the .deps.json. Depending on a situation it might be just a declaration like
```
"Dependency/2.0.1" {},
```
Or it may actually list the assets as well (would contain the actuall .dll path)
But I agree that this doesn't feel like first class support.

swaroop-sridhar commented on docs/dynamic-component-loading.md:
````diff
@@ -0,0 +1,91 @@
+```csharp
+class Assembly
+{
+    public static Assembly LoadFileWithDependencies(string path);
````
@jeffschwMSFT asked about loading assemblies (and their dependencies) to an existing load-context, typically the default load context. If this is a typical use case for developers, we should support it, via an isolation switch.

```csharp
class Assembly
{
    public static Assembly LoadFileWithDependencies(string path, bool Isolated);
}
```
I don't know of any technical reasons not to support this behavior.

jkotas replied:
How would this API handle conflicts? Fail right away if there are any conflicts? Do we have a real app example that would use this API?

jeffschwMSFT replied:
80% of existing use of ALCs load within the default.  I think we should consider this as an append to the end scenario.  If there is a conflict and the existing TPA cannot resolve it, then it would fail to load.  We could either do that proactively or fail at the point of the load of the dependent assembly.

vitek-karas replied:
I'm just curious: How did we learn that 80% of ALC use cases are loading into default?
Also - what do you mean by that: That in the Load the ALC's code calls something like Assembly.LoadFrom?
Also - do we know if that's the intended behavior people wanted?

Do you have a sample I can look at - I would be really interested why would somebody do this...

jkotas replied:
> 80% of existing use of ALCs load within the default.

Do you have a real world example that would be a good candidate for using `LoadFileWithDependencies` into default context API?

sdmaclea replied:
> 80% of existing use of ALCs load within the default

Does this really mean "80% of existing use of ALCs load dependencies within the default ALC"

This would also make them fail to unload....

jeffschwMSFT replied:
If you do a search for ALC in GitHub or nuget you will find that AssemblyLoadContext.Default is the primary use case.  In addition extrapolating from how AppDomains are used (or more often not) leads me to believe that the non-isolated case is likely the primary case.

jeffschwMSFT replied:
For me in reading the existing issues in this space, the #1 issue is people are having problems doing dynamic loads.  Unload ability and isolation are secondary issues.

> Do you have a real world example that would be a good candidate for using LoadFileWithDependencies into default context API?

Most any dynamic load scenario.

vitek-karas replied:
Looking through the various samples and use cases of default ALC - to me it seems the default ALC usage is used a fallback. In this case explicit fallback (where returning null would do the job as well).
I absolutely agree with the fact that # 1 problem is dynamic loads are too hard. But I would argue whether people explicitly want to load into Default.
They probably just want it to work. Using the default is likely a result of debugging, where it failed to load some basic dependency (something from a framework or so), and so the code redirects to default which already has that.

The problem I see with allowing a .deps.json based resolver to load directly into Default is that it is VERY likely to fail. Basically once the plugin has any assembly with the same name as the app, it will fail (even the same exact version, this is because of the current binder behavior which we can't really change). This doesn't feel like a "pit of success" - it seems to be quite the opposite.

Back to the "dynamic load is hard": I think there are 2 main reasons why it's hard:
* If there are dependencies, there's no built-in way to resolve them (other then Assembly.LoadFrom) - this has little to do with loading into default, it's just a lack of functionality.
* Avoiding the first problem can be done by directly loading into Default which has all the dependencies, but now comes the compatibility problem, where it's really easy to fail to load due to version (or other) mismatches. - this is about isolation. We can't solve the collisions without isolation. If we would, we would go back to desktop which allowed resolving higher version reference to a lower version instance - which is really bad.

I haven't yet seen a scenario where loading directly into default with all of the dependency tree is desirable. But there definitely might be one. I'm not against adding an option to do so, but I think it should not be the default. By default we should load with isolation on.

jkotas replied:
> search for ALC in GitHub or nuget you will find that AssemblyLoadContext.Default

I have looked through first several pages of https://github.com/search?l=C%23&q=AssemblyLoadContext.Default&type=Code . I have not found a case that would be a good candidate for loading with isolation off.

jeffschwMSFT replied:
Leaving the default decision aside - I think that customer feedback can help us resolve that.  I do believe there are a lot of cases where customers would rather just have all their code in the same context.  It makes code sharing doable.  As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange.  I have seen very few examples of people getting that right.  Once we have a reasonable higher level communication strategy designed, I could see a path ahead that enabled our customers to get isolation (which I agree is desirable) and allow for sharing in a reasonable way.

swaroop-sridhar replied:
Loading binaries to the parent context is our basic unit of sharing. So, some apps may want to load dependencies (or dependent frameworks) into their own context instead of creating an ALC just to act as a sharing layer [Is this operation significant enough to impact performance?]

Similar behavior can be achieved using existing mechanisms, via a cooperative protocol with the plugin. But LoadWithDependencies() call is more straightforward.

I don't have a strong preference for the isolation switch, we'll only know based on use cases.

swaroop-sridhar replied:
> How would this API handle conflicts? Fail right away if there are any conflicts? 

Yes, the behavior should be identical to loading different versions of the same DLL into an ALC, so I think the load operation should fail right away.

swaroop-sridhar replied:
We should probably call the switch something other than "isolation" because loads in the plugin ALC can cause things to be loaded into the parent ALC, which can affect other plugin ALCs.

jkotas replied:
>  I think that customer feedback can help us resolve that

Agree. It would be very useful to get the first iteration on these APIs (the set that we all agree about) into the first .NET Core 3.0 preview to start getting feedback as soon as possible.

vitek-karas replied:
> As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange. 

This is not the case. The component can choose to share assembly with the app, by not including it in its deployment. There's tooling for this on VS side already (CopyLocal setting for assembly references, similar exists for project and Nuget references). CLI doesn't have this as an option … yet. We might want to add it if we think there's a need. Otherwise modifying the proj file also works.

For example tasks for MSBuild. The plugin (task) implements the `ITask` interface which comes from `Microsoft.Build.Framework.dll`. It makes perfect sense to not include the `Microsoft.Build.Framework.dll` with the plugin, as that's clearly part of the host (MSBuild). Doing that will also inherently share all the types from that assembly between the host and all the plugins.

jeffschwMSFT replied:
WRT customer feedback.  I hope there are multiple avenues to get early feedback: let's move this discussion to coreclr and invite people with existing feedback to chime in, talk with Richard to get early enterprise'ish usage, and get them into preview for feedback.

jeffschwMSFT replied:
>>As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange.

>This is not the case. The component can choose to share assembly with the app, by not including it in its deployment. 

Perhaps we are having a difference in words only... but for me CreateInstance is not a natural pattern and fits within my reflection use case.  If we had a LoadAndUnwrap then perhaps that is pretty natural, as that does not require any prior knowledge of what is loaded where and you get an 'object' back.

AaronRobinsonMSFT replied:
My 2 cents here is that isolation should be the default since that seems to be the entire point of this feature - but perhaps not on this API. I am generally leery of APIs that take `bool` arguments because any API that takes a `bool` is invariably trying to make a decision about what the default behavior. In this case adding a new function that indicates `Assembly.LoadIsolated()` makes way more sense to me and is clear about the very new semantics we are providing.

jeffschwMSFT replied:
+1 on this API in essence being the 'append to an existing ALC' by loading with dependencies and the ALC API being the way to load with isolation.  The more general scenario that I hope we are able to enable is loading multiple plug-ins into the same ALC (be it either default or a new one).  Having a 1:1 relationship of a deps to a ALC (which is what isolation by default in essence turns into) I feel is too limiting.

jkotas replied:
I am wondering whether it would be better to drop the simple convenience API from the first phase given that we do not have clear answer for the right default. We can always introduce it later once we have better data about what the default should be and we have somebody asking for it.

jkotas commented on docs/dynamic-component-loading.md:
```diff
@@ -69,10 +69,19 @@ The second overload adds the ability to specify:
+    * Imitate app behavior exactly - `.deps.json` only provides list of resource probing paths. ALC would then try to find the `<culture>/AssemblyName.dll` in each probing path and resolve the first match.
+    * Use full paths - `.deps.json` resolution actually internally produces a list of full file paths and then trims it to just probing paths. The full file paths could be used by the ALC in a very similar manner to code assemblies.
+* Native libraries - To integrate well the ALC would only get list of native probing paths from the `.deps.json`. It would then use new API to load native library given a probing path and a simple name. Internally this would call into the existing runtime behavior (which tries various prefix/suffix combinations and so on).
+* R2R Images (NI images) - The same list of full paths as for managed assemblies may contain NI images (recognized by file extension). NI images take priority over IL images always. ALC will load the NI image.
```
We are not using the `.ni.dll` suffix for native images. I think it was always the case in .NET Core. There are historic places that still use it, but I do not think we are shipping with the `.ni.dll` extension anywhere.

The problem with using `.ni.dll` suffix for native images is that everything has to be aware of it and strip/add it as necessary.

vitek-karas replied:
I was looking into the runtime. The TPA binder will look for `.ni.dll`, `.ni.exe` and `.ni.winmd` and prioritize those over non-NI paths. Also looking into `AssemblyBinder::GetAssembly` it does make a lot of different decisions if the assembly is NI or not. And as far as I can tell, it only determines that by the `fExplicitBindToNativeImage` parameter which is set to TRUE based on the file extensions as mentioned above.

I admit I haven't debugged through it to see exactly what's going on... but it looks different to what you describe. But maybe I'm just reading the code in the wrong way.

jkotas replied:
I agree that there are places that handle the `.ni.dll` extension. I guess I am trying to say that those paths are not exercised and likely buggy because of we are not shipping binaries with `.ni.dll` extension.

vitek-karas replied:
Got it - that simplifies things.

swaroop-sridhar commented on docs/dynamic-component-loading.md:
```diff
@@ -75,12 +75,12 @@ The second overload adds the ability to specify:
 What happens with various asset types once the ALC decides to load it in isolation.
-* Normal managed assembly (code) - The `.deps.json` parsing code will return list of full file paths. The ALC will just find it there and load it.
+* Normal managed assembly (code) - The `.deps.json` parsing code will return list of full file paths. The ALC will just find it there and load it.  
+Note that R2R images are handled by this as well since they are basically just a slightly different managed assembly.
```
We should mention that if the same R2R image is loaded by two ALCs, only one of them can use the native code. This is not a limitation of the plugin model per say, but we should mention it in the context.

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+## Problem statement
+.NET Core runtime has only limited support for dynamically loading assemblies which have additional dependencies or possibly collide with the app in any way. Out of the box only these scenarios really work:
+* Assemblies which have no additional dependencies other than those found in the app itself (`Assembly.Load`)
+* Assemblies which have additional dependencies in the same folder and which don't collide with anything in the app itself (`Assembly.LoadFrom`)
```
> collide with anything in the app itself

This is an ambiguous statement. Unsure about how to make it more precise, but what this means is definitely different based on the experience of the person answering a question about `Assembly.LoadFrom()`.

vitek-karas replied:
This section is a very short description of existing APIs, it's not meant to be 100% precise. If you can point out what you mean - I honestly don't see anything wrong with the sentence...

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial set of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.
```
>In lot of these cases

"In most of these cases"

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
```
Please include a link to @janvorli 's assembly unload document.

vitek-karas replied:
No such document exists - yet.

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+In its core this is similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
````
> In its core

"At its core"

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+In its core this is similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
````
>like `Assembly.LoadFile` it provides isolation, but also for the dependencies.

That already provides isolation? I thought that is what we are providing here.

vitek-karas replied:
`Assembly.LoadFile` already loads the specified assembly in isolation. But it's very simplistic. It always creates a new ALC every time its called and loads the specified file into it. It doesn't provide any resolution logic and falls back to default for everything. So the loaded assembly itself is fully isolated, but all its dependencies will be resolved from Default.

That's why the new proposed API (if we go with isolation by default) is closest in behavior for `Assembly.LoadFile` (and thus the name as well)

jkotas replied:
> It always creates a new ALC every time its called

Nit: If you call it twice with the same argument, only one ALC gets created. It creates a new ALC per absolute assembly path, not every time it's called.

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
````diff
@@ -0,0 +1,117 @@
+```csharp
+class AssemblyLoadContext
+{
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(string assemblyPath);
````
I see little value for the plain `string` argument. We should have a binary path for this kind of thing either (a) the caller wants predictable and well-defined behavior or (b) the caller has some special scenario and they need to do some work. I think we are providing too many entry points for an already niche scenario in the grand scheme of things. We can _always_ add and make simpler, but removing is impossible.

vitek-karas replied:
Makes sense

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
```
Can you describe the "host" components? Does custom hosting require providing an implementation of this mechanism?

swaroop-sridhar replied:
I guess the host-components here is the code in hostpolicy.dll that does `deps.json` parsing. If a custom host should call `LoadWithDependencies` then it must provide an implementation of this functionality. We should consider whether it is worth extracting the deps.json parsing functionality to a separate DLL to share with other hosting scenarios.

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+    * On the downside, this doesn't provide implicit sharing. Typically only framework assemblies would be shared in this scenario. Component would have to explicitly choose which assemblies to share, by not carrying them with it (this can be setup in project file by using `CopyLocal=false` for assembly references, similar option exists for project and NuGet references as well). This means that t ...
+* **Always load into default** - in this case all loads are done to the parent (Default) load context. In the extreme case a new load context is not really needed. Alternative could be that the load is attempted to the default load context (first by name, then by resolved file path). If that fails (should really only happen in case of a collision), the dependency would be loaded into the new load ...
+    * Inherently shares as much as possible - avoids problems of sharing types used for communication.
+    * Downside is that loading a component will "pollute" the default load context with assemblies form the component.
```
>form the component

"from"

AaronRobinsonMSFT commented on docs/dynamic-component-loading.md:
```diff
@@ -0,0 +1,117 @@
+    * Use full paths - `.deps.json` resolution actually internally produces a list of full file paths and then trims it to just probing paths. The full file paths could be used by the ALC in a very similar manner to code assemblies.
+* Native libraries - To integrate well the ALC would only get list of native probing paths from the `.deps.json`. It would then use new API to load native library given a probing path and a simple name. Internally this would call into the existing runtime behavior (which tries various prefix/suffix combinations and so on).
+
+## Isolation strategies
```
I like these strategy options - smells like an `enum` instead of a `bool` to me. Could also be a `LoadIsolationPolicy` struct.

swaroop-sridhar replied:
Are you thinking of different kinds of isolation settings? For example:
* Load everything in the new ALC
* Load everything in the parent/default ALC (in which case, there is no point creating a new ALC)
* Load in the new ALC only if the parent cannot load it
* Load in the new ALC only if the parent cannot load it without using roll-forward
* ...

I think it is best to enable and leave these isolation settings to the developer, rather than providing API support for all of them -- at least to start with.

AaronRobinsonMSFT replied:
Well the strategies are implying settings, not me. My point here is that a `bool` by its very nature is a binary idea, but based on all these talks it isn't a binary idea rather it is a spectrum. This implies a `bool` is not the appropriate idiom to use for isolation because it isn't "yes or no", but rather "isolate like this ...". An `enum` is one option and a "policy" type struct that indicate how to isolate is another. We are getting into the very concern you and I spoke about earlier this week which is why I am advocating for a very simply scenario that covers a majority of cases and then if a scenario exists that is distinct from the default the developer or consumer of this API will have to answer a series of questions and make real decisions.

> rather than providing API support for all of them

That is a very nuanced statement. Yes and no. Yes, we shouldn't expose a plethora of options that aren't requested, but no we need to think about an API that can grow and adapt. `enum`s and policy structs are the appropriate mechanisms for a malleable API surface, `bool`s are not.

vitek-karas replied:
Just curious which bool are you referring to? The public API as proposed above doesn't provide any way to customize the policy. We've been discussing this quite a bit and the current thinking has somewhat moved, but I'm still trying to prototype some of these ideas before writing it all down. The gist of it is:
- Provide some kind of building blocks, so that devs can compose the policy they want
- If we find a sensible default, implement that as a easy-to-use helper (or possibly the entry point into this problem space) - but this has not been agreed on yet.

vitek-karas commented:
I've started the official review of this design here: dotnet/designs#47
I'm happy to continue some discussions here if you want to.

vitek-karas commented:
This has been superceded by the `AssemblyDependencyResolver` work for the most part.
